models/siegert_neuron.cpp

```diff
@@ -0,0 +1,479 @@
+  updateValue< double >( d, names::tau, tau_ );
+  updateValue< double >( d, names::tau_m, tau_m_ );
+  updateValue< double >( d, names::tau_syn, tau_syn_ );
+  updateValue< double >( d, names::t_ref, t_ref_ );
+
+  if ( V_reset_ >= theta_ )
+    throw BadProperty( "Reset potential must be smaller than threshold." );
+
+  if ( t_ref_ < 0 )
+    throw BadProperty( "Refractory time must not be negative." );
+
+  if ( tau_ <= 0 )
+    throw BadProperty( "time constant must be > 0." );
+
+  if ( tau_m_ <= 0 )
+    throw BadProperty( "Membrane time constant must be > 0." );
+
+  if ( tau_syn_ < 0 )
+    throw BadProperty( "Membrane time constant must not be negative." );
+}
+
+void
+nest::siegert_neuron::State_::get( DictionaryDatum& d ) const
+{
+  def< double >( d, names::rate, r_ ); // Rate
+}
+
+void
+nest::siegert_neuron::State_::set( const DictionaryDatum& d )
+{
+  updateValue< double >( d, names::rate, r_ ); // Rate
+}
+
+nest::siegert_neuron::Buffers_::Buffers_( siegert_neuron& n )
+  : logger_( n )
+{
+}
+
+nest::siegert_neuron::Buffers_::Buffers_( const Buffers_&, siegert_neuron& n )
+  : logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::siegert_neuron::siegert_neuron()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::siegert_neuron::siegert_neuron( const siegert_neuron& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::siegert_neuron::init_state_( const Node& proto )
+{
+  const siegert_neuron& pr = downcast< siegert_neuron >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::siegert_neuron::init_buffers_()
+{
+  // resize buffers
+  const size_t quantity = kernel().connection_manager.get_min_delay();
+  B_.drift_input_.resize( quantity, 0.0 );
+  B_.diffusion_input_.resize( quantity, 0.0 );
+  B_.last_y_values.resize( quantity, 0.0 );
+
+  B_.logger_.reset(); // includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::siegert_neuron::calibrate()
+{
+  B_.logger_
+    .init(); // ensures initialization in case mm connected after Simulate
+
+  const double h = Time::get_resolution().get_ms();
+
+  // propagators
+  V_.P1_ = std::exp( -h / P_.tau_ );
+  V_.P2_ = -numerics::expm1( -h / P_.tau_ );
+}
+
+/* ----------------------------------------------------------------
+ * Update and event handling functions
+ */
+
+bool
+nest::siegert_neuron::update_( Time const& origin,
+  const long from,
+  const long to,
+  const bool wfr_update )
+{
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  bool done = true;
```

Comment:
@heplesser I agree, that this is easier to understand. I changed `done` to `wfr_tol_exceeded` in all models. I do however still return `done` (meaning `not wfr_tol_exceeded`) from `wfr_update` since in `event_delivery_manager` and the `simulation_manager` I like `done` much better than `wfr_tol_exceeded`.